_posts/2022-10-09-decentralization-of-zk-rollups.md

```diff
@@ -0,0 +1,192 @@
+
+All current zk rollups are centralized. Some do not even have a projected path to decentralization. This is not a good situation to be in. Some features in a truly decentralized zk-rollup which we hope to achieve:
+
+
+
+1. Safety / Soundness. Ability to resist a malicious party from taking control over the network, and dictating the present and future state of the network. 
+2. Censorship resistance. Everyone should be able to have access to the network and interact with other participants. The network should not be able to deny this right to any individual. It should also be possible for users to withdraw their funds back to L1 or force execute transactions in L2 by paying gas in L1, even if the network is compromised, or the operators are limiting user activities on the network. 
+3. Low node / network participant requirements. Individuals should be able to build and run their own node, prover or sequencer without an unreasonable amount of difficulty. Expecting state of the art or highly specialized equipment and high compute resources should be avoided. 
+4. Economically aligned incentives. Participants should be rewarded equally for upkeep of the network, in order to encourage new participants to enter. 
+5. Liveness. Honest messages should always be included and available to block producers. Chances of halting or lapses in network availability  should be mitigated. 
+
+
+## Upgrade Keys
+
+Upgrade keys are public keys that have been whitelisted by a rollup’s L1 smart contract, and they have two main benefits:
+
+
+
+1. Can fix buggy code, an eventuality that cannot be escaped, as reducing the chance of this happening is very hard. 
+2. Can improve the protocol to update with new technology advancements. This is available on L1s through hard forks.
+
+But, this amount of centralized control over an L2 is not very safe as this entity could change the complete state and protocol. 
+
+How do we mitigate this risk? We have identified some ways to move forward:
+
+Abolish upgrade keys permanently. If there are any changes to be made with the original smart contract, bundle them together and deploy a new smart contract with the changes. This is extremely abrupt and harsh. You will then have to entice the miners and users to organically switch over from the “old” network. You can ease into it by airdropping tokens of the “new” rollup to users of the “old” rollup equivalent to their last checked balance, or by accepting the wrapped L1 tokens from the previous L2 rollup.
```

Comment:
This is not practical, but there are interesting case studies such as Loopring, which had at least three versions of their "layer 2" and I recall they have tried this mechanism in the past (need to verify). It should be coupled with some real world examples and references